dashboard/test/ui/features/star_labs/applab/shared_apps.feature

```diff
@@ -32,7 +32,7 @@ Feature: App Lab Scenarios
     And I wait until element "#divApplab > .screen > button#testButton1" is visible
     Then element "#testButton1" contains text "Click me"
     When I press "testButton1"
-    Then element "#testButton1" contains text "Clicked"
+    And I wait until element "#testButton1" contains text "Clicked"
```

Comment:
Out of curiosity would [this](https://github.com/code-dot-org/code-dot-org/blob/9bf8dd75548bedbeb3aac09402536e8db4661302/dashboard/test/ui/features/step_definitions/steps.rb#L705) also have worked?

Reply:
I tried this command and it works as well.

Is there a reason to use one command vs the other?

Also, there are different versions:
- `When I press "testButton1"
    Then element "#testButton1" eventually contains text "Clicked"`
- `And I press "testButton1"
    And element "#testButton1" eventually contains text "Clicked"`

It doesn't seem to make a difference to use when/then vs and/and. Is there a convention? Thanks!

Reply:
It looks like "wait until element X contains text Y" uses jQuery, [here](https://github.com/code-dot-org/code-dot-org/blob/79e0ce080e16e7d9bf9f0f523cb70536ce6d26a6/dashboard/test/ui/features/step_definitions/steps.rb#L200-L202), injected into the browser, while "element X eventually contains text Y" uses Selenium functionality and also ensures that the element includes Y but isn't an necessarily an exact match, [here](https://github.com/code-dot-org/code-dot-org/blob/79e0ce080e16e7d9bf9f0f523cb70536ce6d26a6/dashboard/test/ui/features/step_definitions/steps.rb#L705-L707) and [here](https://github.com/code-dot-org/code-dot-org/blob/79e0ce080e16e7d9bf9f0f523cb70536ce6d26a6/dashboard/test/ui/features/support/browser_helpers.rb#L61-L65).  So I guess you can pick the best one for the job, though it's probably just a quirk of history that we have two similar but different steps.

I believe those keywords are interchangeable.  Details [here](https://cucumber.io/docs/gherkin/reference/).

Reply:
Thank you for the refs and the resource link! Definitely will take another look to make sure that the new step I'm adding is not duplicating an existing step.

Reply:
I will go ahead and use "wait until element X contains text Y" because it is used four other times in this file.

Also, I looked for an existing step command which waited a given amount of time before checking the value of an element, but I did not find it. I did see 'Then I wait up to X seconds for element Y to have css property Z equal to W". But not a step that checked the value of an element within a certain time interval.